### PR TITLE
Espressif ESP32 Hardware Acceleration Update and Cleanup

### PR DESCRIPTION
**Pull request by gojimmypi**
# Description

This PR updates the Espressif hardware acceleration code, mostly surrounding SHA2.

There's a new `ctx.lockDepth` that keeps track of RTOS calls that lock the hardware for exclusive use. This was implemented as the Espressif lib does not properly track `ref_counts[periph]` in [periph_ctrl.c](https://github.com/espressif/esp-idf/blob/release/v4.4/components/driver/periph_ctrl.c), specifically [periph_module_enable](https://github.com/espressif/esp-idf/blob/c9140caf8cb501f17e4e0703a43a5693a6c28f6d/components/driver/periph_ctrl.c#L15) and [periph_module_disable](https://github.com/espressif/esp-idf/blob/c9140caf8cb501f17e4e0703a43a5693a6c28f6d/components/driver/periph_ctrl.c#L26). If/when RTOS unexpected failures occur, there's a new `esp_unroll_sha_module_enable` to facilitate returning the peripheral to a known enabled state for the given `ctx` object. 

The `esp_sha_hw_lock` and `esp_sha_hw_unlock` now take a `ctx` parameter.

Instead of assuming exact location in memory space (e.g. for hardware support on newer chipsets that may rearrange things, particularly on architecture change between Xtensa and RISC-V) - there's now a `switch` statement with reference to HW memory macros instead of a hard-coded address parameter value. For example, see `esp_sha_start_process` using `sha->sha_type`

Some preliminary `CMakeFile` changes were implemented to address some of the breaking changes between ESP-IDF V4 and the new V5.x . Additional work is needed to fully support V5.

Various other code cleanup, formatting, and commenting.

This is an interim PR with additional changes and additions planned with other hardware acceleration.

Fixes zd#  (n/a)

# Testing

Confirmed to be working with latest SSH to UART project. There should probably be more robust tests created in a future PR to actually compare HW & SW results and report success/failure.

# Checklist

 - [ ] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

File: IDE/Espressif/ESP-IDF/libs/component.mk [hunk 1]
```diff
@@ -1,17 +1,37 @@
...
-CFLAGS +=-DWOLFSSL_USER_SETTINGS
-
-COMPONENT_OBJEXCLUDE := wolfcrypt/src/aes_asm.o
-COMPONENT_OBJEXCLUDE += wolfcrypt/src/evp.o
-COMPONENT_OBJEXCLUDE += wolfcrypt/src/misc.o
-COMPONENT_OBJEXCLUDE += src/bio.o
+#
+#  Copyright (C) 2014-2022 wolfSSL Inc.
+#
+# This file is part of wolfSSH.
```
Comment (dgarske) [inline IDE/Espressif/ESP-IDF/libs/component.mk:4]: wolfSSL license header.

File: wolfcrypt/src/sha256.c [hunk 2]
```diff
@@ -1861,13 +1898,20 @@ int wc_Sha256GetHash(wc_Sha256* sha256, byte* hash)
...
+    {
         sha256->ctx.mode = ESP32_SHA_SW;
+    }
 #endif
 
         wc_Sha256Free(&tmpSha256);
     }
+
+#if defined(WOLFSSL_USE_ESP32WROOM32_CRYPT_HASH_HW)
+    /* TODO: Make sure the ESP32 crypto mutex is released (in case final is not
```
Comment (dgarske) [inline wolfcrypt/src/sha256.c:1912]: Should we fix this TODO? Should be able to detect and release in the `wc_Sha256Free`

File: wolfssl/wolfcrypt/port/Espressif/esp32-crypt.h [hunk 3]
```diff
@@ -22,39 +22,44 @@
 
 #define __ESP32_CRYPT_H__
 
+#ifdef WOLFSSL_USER_SETTINGS
```
Comment (dgarske) [inline wolfssl/wolfcrypt/port/Espressif/esp32-crypt.h:25]: Please include wolfssl/wolfcrypt/settings.h always instead. It does this logic at top and ensures build options checks are done.

**Review by dgarske — CHANGES_REQUESTED** [review]
Note we are in release code freeze. Please coordinate with Jacob is you want this in the release... it might be possible.

File: IDE/Espressif/ESP-IDF/libs/CMakeLists.txt [hunk 4]
```diff
@@ -1,6 +1,24 @@
 #
+#  Copyright (C) 2014-2022 wolfSSL Inc.
+#
+# This file is part of wolfSSH.
```
[the remark this answers is not in the document]
Comment (gojimmypi) [reply IDE/Espressif/ESP-IDF/libs/CMakeLists.txt]: good catch! updated.

File: wolfcrypt/src/sha256.c (hunk 2 above)
Comment (gojimmypi) [reply wolfcrypt/src/sha256.c:1931]: I added a check of `sha256->ctx.lockDepth` in `wc_Sha256Free` to ensure the mutex is actually released.

File: wolfssl/wolfcrypt/port/Espressif/esp32-crypt.h (hunk 3 above)
Comment (gojimmypi) [reply wolfssl/wolfcrypt/port/Espressif/esp32-crypt.h]: that's good to know, thanks. fixed.

**Review by dgarske — APPROVED** [review]

**Pull request merged**